### ISSUE TITLE
[explicit-module-boundary-types] make rule apply when using function wrappers

### ISSUE THREAD

**Issue by NicBright** [typescript-eslint/typescript-eslint#4486]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/explicit-module-boundary-types": "error"
  }
}
```

```TS
function memoizeFactory<T extends (...args) => any>(arg: T): T {
    let memoized = arg; // assume we have memoized arg somehow
    return memoized; // memoized has the same signature as arg
}

export const myMemoizedFactory = memoizeFactory(() => {
    return 5;
});

export const unmemoizedFactory = () => {
    return 5;
};
```

**Expected Result**

The rule should add an error to the line exporting myMemoizedFactory, because the supplied factory function to the function wrapper is insufficiently typed.

**Actual Result**

![image](https://user-images.githubusercontent.com/1225659/150767921-ffe349f2-5a43-4b62-bb68-0920d5ecce14.png)

The error is only added to the "unmemoizedFactory" export (this demonstrates that the rule works as expected when the same factory function is not wrapped.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `5.10.0` |
| `@typescript-eslint/parser`        | `5.10.0` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.32.0` |
| `node`                             | `14.17.0` |

**Comment by bradzacher** [comment]
The rule doesn't do this on purpose.

It's impossible to tell via the AST exactly what a function does. So the rule has no idea that `memoizeFactory` simply "wraps" the function.
It's easy for a human to tell - but hard for a computer to tell.
So we have two choices: assume the function is passed into a function wrapper, or don't. The choice is clear to not because in JS the callback pattern is very prevalent - much more than the function wrapper pattern.

Additionally, the rule is designed similarly to `explicit-function-return-type` (or rather - the same as - they use the same code). It ignores "typed function expressions" - functions where we assume the type is easily inferrable based on the usage, like a function in an argument position.
This behaviour is configurable: https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/explicit-module-boundary-types.md#allowtypedfunctionexpressions

**Comment by NicBright** [comment]
Hi @bradzacher 

Thanks for your quick response to my suggestion. You're right: in general a computer can't tell what a function does without looking at the implementation. In this case, however, I think it's not necessary. What could be done is looking at the signature of the wrapper function. In my example, it's:

```typescript
function memoizeFactory<T extends (...args) => any>(arg: T): T {
    let memoized = arg; // assume we have memoized arg somehow
    return memoized; // memoized has the same signature as arg
}
```

I suggest the following algorithm:

1. Look at the return type of the function (`memoizeFactory`). If it's a generic type variable, then continue, else we're done
2. Look at the generic constraint of the return type variable. If it can be boiled down to a function type, then continue, else we're done
3. Foreach argument of the argument list of the function (`memoizeFactory`), check:
  3.1. has the argument the same type as the return type of the function (i.e. the same generic type variable)? Then treat it like a regular function to be checked with `explicit-module-boundary-types`. Else, continue with the next argument.
4. we're done

In my case, the algorithm would spot that `(arg: T)` is an argument which type is returned unmodified and thus the rule could know that the argument needs to be checked.

**Comment by bradzacher** [comment]
> What could be done is looking at the signature of the wrapper function

Unfortunately the rule doesn't access type information. Adding it comes with a large cost and is a breaking change. For various reasons it locks a lot of users out of using the rule - so we don't do it unless there is a *very* good reason to do so.

Which means this rule can only act based on the AST information.

**Comment by NicBright** [comment]
I didn't know that. I just assumed it had type info because it's checking the type but apparently some parts of the type are accessible from the AST as it seems (those parts that are explicitly written in the source code).

One last possibility comes to my mind:

What about adding a config option to specify commonly used wrapper functions whose argument(s) should be checked as if they weren't wrapped? It could be named `transparentWrapperFunctions` and described something like this:

```
/**
   * An array of function/method names that return their argument unmodified signature-wise, and can therefore be considered being transparent. The rule will then consider the argument as if it wasn't wrapped at all. One such example would be wrapping a function with a 'memoize()' method.
   */
```
